### ISSUE TITLE
formatOnType does not respect current line indendtiation

### ISSUE THREAD

**Issue by dbaeumer** [Microsoft/TypeScript#6700]
_From @isidorn on January 26, 2016 15:50_

1. js workspace, have salsa enabled
2. "editor.formatOnType": true
3. The format on type is too aggrressive and does not respect current indentiation. This is especially noticable when trying to create object literals
4. if salse is disabled this works nicely

_Copied from original issue: Microsoft/vscode#2377_

**Comment by dbaeumer** [comment]
@isidorn I can't see any weirdness. Can you please provide some steps. I tested ts and js and they behave the same. Both honoring indent setting.

Please be aware that if a indent of for example 4 spaces exist and you have insert tab and tab size > 4 then format on type from TS will stil insert 4 spaces to align the code correctly.

**Comment by dbaeumer** [comment]
_From @isidorn on January 28, 2016 15:23_

```
var t = {
  name: isidor,
// this line gets indented all the way to the start once I press enter after name property
}
```

**Comment by dbaeumer** [comment]
Same happens for TS.

**Comment by dbaeumer** [comment]
I debugged it and this is a tsserver behavior. When pressing enter after Isidor, VSCode inserts the right indention but the tsserver returns and edit that deletes that indention. IMO the tsserver should be smart and if Isidor is ended with , adjust the indention since there is a high likelihood that the user inserts another property.

**Comment by billti** [comment]
Sounds like this is a general issue with TypeScript also, not a Salsa issue.

**Comment by mlakmal** [comment]
Does this also remove the space after function keyword like below? Our lint rules require space after function. But after latest vs code 0.10.10 it removes spaces on reformatting ...

var test = function() { };

**Comment by zhengbli** [comment]
@mlakmal This issue is referring to the "format on enter" feature; the issue you are referring to is different. Though it may worth opening another issue for tracking.

**Comment by johnwu5** [comment]
Another related issue, when adding comments as below example:
![screen](https://cloud.githubusercontent.com/assets/1189629/13655545/0490ff74-e615-11e5-9175-e64c644b812b.png)

Prior to 0.10.10 a space is added after the * in next new line when pressing Enter. In 0.10.10, the space is removed, ALT-Z will bring it back.

**Comment by mlakmal** [comment]
@egamma is the issue 3796 in vs code related to this issue...

**Comment by mhegazy** [comment]
The formatter never removed spaces from comments. i think this is the editor doing this. @egamma any thoughts?

**Comment by jrieken** [comment]
@mhegazy Is there any progress on this? We have multiple customers complaining about format on typing being usable.

**Comment by mhegazy** [comment]
@jrieken do you have more information about this issue? are there other cases? i can not reproduce the original issue.

**Comment by 4kochi** [comment]
See https://github.com/Microsoft/vscode/issues/3797 and https://github.com/Microsoft/vscode/issues/5656#issuecomment-214190434

**Comment by zhengbli** [comment]
This issue was already fixed by #7307, for some reason I forgot to close this issue. The fix is in the nightly versions now.

**Comment by jrieken** [comment]
@zhengbli Can you please comment on https://github.com/Microsoft/vscode/issues/3797#issuecomment-215330505

**Comment by mhegazy** [comment]
does not look like it is fixed.

**Comment by mhegazy** [comment]
from https://github.com/Microsoft/vscode/issues/3797#issuecomment-215330505
![image](https://cloud.githubusercontent.com/assets/8000722/14895027/791e3eaa-0d2b-11e6-974e-3332569d913a.png)

**Comment by zhengbli** [comment]
@dbaeumer @jrieken 

I am testing a fix, however I found that it may have something to do with VSCode's own formatting behavior.

For example, when I have the `formatOnType` set to `true`, I observed this:
![gif](https://cloud.githubusercontent.com/assets/1171301/15053172/1be38624-12b6-11e6-82df-17a642850d38.gif)

While the VSCode log said:
```
Sending request: change (76). Response expected: no. Current queue length: 0
Arguments: {
    "file": "c:/Users/lizhe/Documents/playground/testLocalFormatting/test.ts",
    "line": 1,
    "offset": 35,
    "endLine": 1,
    "endOffset": 35,
    "insertString": "\r\n    "
}

Sending request: formatonkey (77). Response expected: yes. Current queue length: 0
Arguments: {
    "file": "c:/Users/lizhe/Documents/playground/testLocalFormatting/test.ts",
    "line": 2,
    "offset": 5,
    "key": "\n"
}

Response received: formatonkey (77). Request took 3 ms. Success: true 
Result: [
    {
        "start": {
            "line": 1,
            "offset": 16
        },
        "end": {
            "line": 1,
            "offset": 17
        },
        "newText": ""
    },
    {
        "start": {
            "line": 1,
            "offset": 31
        },
        "end": {
            "line": 1,
            "offset": 32
        },
        "newText": ""
    }
]

Sending request: change (78). Response expected: no. Current queue length: 0
Arguments: {
    "file": "c:/Users/lizhe/Documents/playground/testLocalFormatting/test.ts",
    "line": 2,
    "offset": 1,
    "endLine": 2,
    "endOffset": 5,
    "insertString": ""
}

Sending request: change (79). Response expected: no. Current queue length: 0
Arguments: {
    "file": "c:/Users/lizhe/Documents/playground/testLocalFormatting/test.ts",
    "line": 1,
    "offset": 31,
    "endLine": 1,
    "endOffset": 32,
    "insertString": ""
}

Sending request: change (80). Response expected: no. Current queue length: 0
Arguments: {
    "file": "c:/Users/lizhe/Documents/playground/testLocalFormatting/test.ts",
    "line": 1,
    "offset": 16,
    "endLine": 1,
    "endOffset": 17,
    "insertString": ""
}
```

It looks like the server is returning a correct change, while VSCode decides to delete the spaces on the second line. Is there any way to turn off VSCode's formatting for testing purpose?

**Comment by basarat** [comment]
Indeed the TypeScript language service `getIndentationAtPosition` on that line [gives the right result as I use it in alm](https://github.com/alm-tools/alm/blob/f8d597374437cbc5911f0cabe26124a29d8eaa11/src/app/codemirror/mode/typescriptMode.ts#L201-L214): 

![alm](https://cloud.githubusercontent.com/assets/874898/15061462/2cef8dbe-1379-11e6-82a8-893c1b2d2e73.gif)

But as you can tell from the comments I struggled to get it just right originally because of my misconfigurings and stuff :) :rose:

**Comment by jrieken** [comment]
@alexandrudima Can you comment?

**Comment by 4kochi** [comment]
Any progress on this issue? 

I would also like to mention that the problem only occurs if one hit enter two times. 

![mai-13-2016 09-51-49](https://cloud.githubusercontent.com/assets/1742426/15241468/6270e100-18f0-11e6-9f96-b9df54d9e740.gif)

**Comment by mhegazy** [comment]
Pinging @alexandrudima  and @jrieken

**Comment by dbaeumer** [comment]
I debugged it again today with the latest code alpha build which uses tsserver 1.8.10 and here is what is happening for a TS file with format on type and tab indention configured:

```typescript
function foo(): void {
<tab ch><cursor is here>
}
```

Press enter
You receive the following edits from the tsserver:
```json
Sending request: change (144). Response expected: no. Current queue length: 0
Arguments: {
    "file": "p:/mseng/VSCode/Playgrounds/bugs/TS6700/test.ts",
    "line": 2,
    "offset": 2,
    "endLine": 2,
    "endOffset": 2,
    "insertString": "\r\n\t"
}

Sending request: change (145). Response expected: no. Current queue length: 0
Arguments: {
    "file": "p:/mseng/VSCode/Playgrounds/bugs/TS6700/test.ts",
    "line": 2,
    "offset": 1,
    "endLine": 2,
    "endOffset": 2,
    "insertString": ""
}

Sending request: formatonkey (146). Response expected: yes. Current queue length: 0
Arguments: {
    "file": "p:/mseng/VSCode/Playgrounds/bugs/TS6700/test.ts",
    "line": 3,
    "offset": 2,
    "key": "\n"
}

Response received: formatonkey (146). Request took 6 ms. Success: true 
Result: [
    {
        "start": {
            "line": 3,
            "offset": 1
        },
        "end": {
            "line": 3,
            "offset": 2
        },
        "newText": ""
    }
]

Sending request: change (147). Response expected: no. Current queue length: 0
Arguments: {
    "file": "p:/mseng/VSCode/Playgrounds/bugs/TS6700/test.ts",
    "line": 3,
    "offset": 1,
    "endLine": 3,
    "endOffset": 2,
    "insertString": ""
}
```

The first notification is the enter press with the \t we insert. But then interestingly vscode send a chance request that deletes in indentation on line 2. The a format on type is sent which returns a edit from the tsserver that removes the \t on line 3. 

So the tsserver does remove the \t on line 3. However I am not sure if this is because we remove the \t on line 2 before.

**Comment by dbaeumer** [comment]
I disable trimTrailingWhitespace but same behavior.

**Comment by dbaeumer** [comment]
Exactly the same behavior occurs with JS files.

**Comment by dbaeumer** [comment]
I talked to @alexandrudima and the removal of the <tab> on line 2 is a feature that is controlled via the setting "editor.trimAutoWhitespace". The feature makes absolutely sense since it takes out whitespaces the editor inserted automatically. Also the sequence of the events seems correct since we ask the TS server to format on type on the state of the document like presented in the editor. 

I disabled the setting "editor.trimAutoWhitespace":false and intestingly the formatting is still incorrect. The trace looks like this in the same example in a TS file:

```json
Sending request: change (47). Response expected: no. Current queue length: 0
Arguments: {
    "file": "p:/mseng/VSCode/Playgrounds/bugs/TS6700/test.ts",
    "line": 2,
    "offset": 2,
    "endLine": 2,
    "endOffset": 2,
    "insertString": "\r\n\t"
}

Sending request: formatonkey (48). Response expected: yes. Current queue length: 0
Arguments: {
    "file": "p:/mseng/VSCode/Playgrounds/bugs/TS6700/test.ts",
    "line": 3,
    "offset": 2,
    "key": "\n"
}

Response received: formatonkey (48). Request took 2 ms. Success: true 
Result: [
    {
        "start": {
            "line": 2,
            "offset": 1
        },
        "end": {
            "line": 2,
            "offset": 2
        },
        "newText": ""
    },
    {
        "start": {
            "line": 3,
            "offset": 1
        },
        "end": {
            "line": 3,
            "offset": 2
        },
        "newText": ""
    }
]

Sending request: change (49). Response expected: no. Current queue length: 0
Arguments: {
    "file": "p:/mseng/VSCode/Playgrounds/bugs/TS6700/test.ts",
    "line": 3,
    "offset": 1,
    "endLine": 3,
    "endOffset": 2,
    "insertString": ""
}

Sending request: change (50). Response expected: no. Current queue length: 0
Arguments: {
    "file": "p:/mseng/VSCode/Playgrounds/bugs/TS6700/test.ts",
    "line": 2,
    "offset": 1,
    "endLine": 2,
    "endOffset": 2,
    "insertString": ""
}
```

In this case the TS server sends to edits back to delete the tab character on line 2 and one to delete it on line 3.

Since the tsserver has an AST and therefore knows that the format happens inside a body of a function it should IMO propose an indentation one greater then the one of the function keyword.

@zhengbli @mhegazy can you please comment on how to proceed here.

**Comment by zhengbli** [comment]
I agree the master version does still do the wrong thing for the issue, though I was testing a fix. After I disabled the `trimAutoWhitespace`, now it works as expected using my fix. 

I think what is happening is:

1. The user hits enter, a new line is inserted with spaces added by VSCode
2. VSCode sends the `formatOnType` command to tsserver.
3. Tsserver found that the current line is nothing but space. **Then it doesn't touch the current line to avoid messing the current indentation**, and move the format range to the previous lines, until it found a line with non-space character.
4. Tsserver sends back edits. The VSCode found the current line is not touched, **then it decides to trim auto spaces**
5. The indentation was removed.

The fix is currently at https://github.com/zhengbli/TypeScript/tree/formatOnTypeCurrentLine
Though I am finalizing it soon.

**Comment by dbaeumer** [comment]
@zhengbli can you attach a tsserver trace for this. According to  @alexandrudima VS Code doesn't remove any whitespaces on lines the cursor is on. Neither for remove trailing whitespace on save nor for the trim auto whitespace. So in step 4 in our example above VS Code should not remove and any indentation since that line contains the cursor.

And could you clarify what:

> and move the format range to the previous lines, until it found a line with non-space character.

means.

**Comment by zhengbli** [comment]
The screen recording of the sessions with auto trim spaces enabled / disabled:
![formatontype](https://cloud.githubusercontent.com/assets/1171301/15521782/900abf52-21c2-11e6-9882-a371f47f711d.gif)

The trace log with auto trim enabled:
[trace_trim.txt](https://github.com/Microsoft/TypeScript/files/280636/trace_trim.txt)

The trace log with auto trim disabled:
[trace_no_trim.txt](https://github.com/Microsoft/TypeScript/files/280637/trace_no_trim.txt)

The diffing shows that in the "no_trim" case, VSCode sends an additional `change` which deletes the indentation.

**Comment by zhengbli** [comment]
> and move the format range to the previous lines, until it found a line with non-space character.

To the server essentially `formatOnType` is the same feature as generally formatting the entire document, only that for it the formatting span is only the two lines nearby. To determine where the formatting span ends, we search from the cursor position backwards, and skip all space characters.  Therefore when the current line content is `____|` (`_`indicates space, `|` indicates the cursor), the current line will be skipped, meaning the formatting span ends at the end of last line, so the current line will not be touched.

**Comment by dbaeumer** [comment]
@zhengbli thanks for the case. I opened https://github.com/Microsoft/vscode/issues/6862 since this clearly looks like an editor problem.

However there is still the problem with the two new lines where the traces shows that the tsserver removes the indentation. This is a very annoying bug in VS Code and basically forces us to ask users to disable formatting on type. Are there any plans to have this fixed in 1.8.x ?

**Comment by zhengbli** [comment]
@dbaeumer yes that is a problem. My new fix should take care of that, however I don't think we are going to port that to 1.8. @mhegazy may have comments on the matter.

**Comment by dbaeumer** [comment]
@zhengbli Since this was a major issue for us we added some workaround code in the TS extension. Please ping me when this is fixed in 2.0 so that I can remove the workarounds.

**Comment by zhengbli** [comment]
Sure, thank you!
